app/src/main/java/com/github/geohunt/app/data/repository/bounties/TeamsRepository.kt

```diff
@@ -0,0 +1,137 @@
+package com.github.geohunt.app.data.repository.bounties
+
+import com.github.geohunt.app.data.exceptions.TeamNotFoundException
+import com.github.geohunt.app.data.repository.UserRepositoryInterface
+import com.github.geohunt.app.model.Team
+import com.github.geohunt.app.utility.toMap
+import com.google.firebase.database.*
+import com.google.firebase.database.ktx.snapshots
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.tasks.await
+import kotlinx.coroutines.withContext
+
+/**
+ * Team repository for a specific bounty.
+ */
+class TeamsRepository(
+    bountyReference: DatabaseReference,
+    private val userRepository: UserRepositoryInterface,
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+): TeamsRepositoryInterface {
+    private val teams = bountyReference.child("teams")
+
+    override suspend fun joinTeam(teamId: String, userId: String) {
+        withContext(ioDispatcher) {
+            if (getUserTeam(userId).first() != null) {
+                return@withContext
+            }
+
+            teams.child(teamId).child("members")
+                    .child(userId)
+                    .setValue(true)
+                    .await()
+        }
+    }
+
+
+    override suspend fun leaveTeam(userId: String) {
+        val team = getUserTeam(userId).first() ?: return
+
+        withContext(ioDispatcher) {
+            teams.child(team.teamId).child("members").child(userId).removeValue()
+        }
+    }
+
+    override fun getTeam(teamId: String): Flow<Team> {
+        return teams.child(teamId).snapshots
+                .map { snapshotToTeam(it) }
+                .flowOn(ioDispatcher)
+    }
+
+    override suspend fun getUserTeam(): Flow<Team?> = getUserTeam(userRepository.getCurrentUser().id)
+
+    override fun getTeamScore(team: Team): Flow<Long> =
+        teams.child(team.teamId).child("score").snapshots
+            .map { it.getValue(Long::class.java) ?: throw TeamNotFoundException(team.teamId) }
+            .flowOn(ioDispatcher)
+
+    override fun getUserTeam(userId: String): Flow<Team?> {
+        return getTeams().map {
+            it.firstOrNull { team -> team.membersUid.contains(userId) }
+        }
+    }
+
+    override fun getTeams(): Flow<List<Team>> {
+        return teams.snapshots
+                .map { it -> it.children.map { snapshotToTeam(it) } }
+                .flowOn(ioDispatcher)
+    }
+
+    override suspend fun createTeam(name: String): Team {
+        return createTeam(name, userRepository.getCurrentUser().id)
+    }
+
+    override suspend fun createTeam(name: String, teamLeaderUid: String): Team {
```

Comment:
This function could be private since we don't want people to create teams with someone else's uid.

Reply:
That's going to fuck up my tests.